keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -347,6 +349,30 @@ correct order.
 [enables]:https://github.com/kubernetes/kubernetes/blob/220498b83af8b5cbf8c1c1a012b64c956d3ebf9b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L703
 [filter]:#updating-storageversion
 
+#### Agreements
+
+1. Storageversion updates will be triggered by CRD create and update watch events
+2. We will perform async updates of storageversions such that we do not block writes of other unrelated APIs(for ex, crdA writes waiting for update of crdB's storageversion). These requests should be handled asynchronously
+3. We will block a storageversion update until the teardown of prev storage is finished
+   1. this means we will wait to update storageversion of a CRD until
+      1. old storage of the same CRD is deleted
+      2. pending CR writes using old storageversion of the same CRD are completed
+   2. We will do this to prevent publishing a newer storageversion while a pending CR write is still in progress. Otherwise, if the pending CR write finishes writing the CR using an old version, the storageversion API would not reflect that and the object will remain in etcd, encoded in an old version forever 
+4. We will block new CR writes for a CRD until we have published its latest storageversion. This is discussed more in the [limitations] section
+
+[limitations]:#Limitations 
+
+#### Limitations
+
+When a storageversion of a CRD is updated, we will ensure that all new CR writes:
+1. wait for the latest storageversion to be published
+1. wait for the CRD handler to be replaced
+
+This implies that we will take a write outage for the duration of the storageversion update. We will allow this for the following reasons:
+1. blocking CR writes till the SV update is finished is in-line with how we handle requests for built-in resources
```

Comment:
This is slightly different than built-ins right? Built-in blocking can only occur once on apiserver startup while CR writes can be blocked every time a CRD is updated? Nothing wrong with maintaining the same approach, just something to consider.

Reply:
That's right, for CRDs - we will always block new CR writes until  the latest SV update is finished. Allowing CR writes while the SV update is pending can cause problems like the old handler not being able to understand the new CRD version that may be persisted in etcd by another server.